gobblin-runtime/src/test/java/org/apache/gobblin/runtime/messaging/data/FileSystemMessageBufferTest.java

```diff
@@ -0,0 +1,129 @@
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.mockito.Mockito;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.testng.Assert;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import com.google.common.io.ByteStreams;
+import com.google.common.io.Files;
+import com.typesafe.config.Config;
+import com.typesafe.config.ConfigFactory;
+import com.typesafe.config.ConfigValueFactory;
+
+import org.apache.gobblin.runtime.messaging.DynamicWorkUnitConsumer;
+import org.apache.gobblin.runtime.messaging.hdfs.FileSystemMessageBuffer;
+
+import static org.apache.gobblin.runtime.messaging.DynamicWorkUnitConfigKeys.DYNAMIC_WORKUNIT_HDFS_PATH;
+import static org.apache.gobblin.runtime.messaging.DynamicWorkUnitConfigKeys.DYNAMIC_WORKUNIT_HDFS_POLLING_RATE;
+import static org.mockito.Matchers.any;
+import static org.mockito.Mockito.times;
+
+
+public class FileSystemMessageBufferTest {
+  public final static Logger LOG = LoggerFactory.getLogger(FileSystemMessageBufferTest.class);
+
+  FileSystemMessageBuffer messageBuffer;
+  FileSystem fs;
+  Path path;
+
+  @BeforeClass
+  public void testConstructBuffer()
+      throws IOException {
+    LOG.info("log something");
+    File dir = Files.createTempDir();
+    this.fs = FileSystem.get(dir.toURI(), new Configuration());
+    Config cfg = ConfigFactory.empty()
+        .withValue(DYNAMIC_WORKUNIT_HDFS_PATH, ConfigValueFactory.fromAnyRef(dir.getAbsolutePath()))
+        .withValue(DYNAMIC_WORKUNIT_HDFS_POLLING_RATE, ConfigValueFactory.fromAnyRef(5000));
+    FileSystemMessageBuffer.Factory factory = new FileSystemMessageBuffer.Factory(cfg);
+    messageBuffer = (FileSystemMessageBuffer) factory.getBuffer("HDFS_channel");
+    path = messageBuffer.getPath();
+
+  }
+
+  @Test
+  public void testPublish() throws IOException {
+    String workUnitId = "workUnit0";
+    List<String> laggingPartitions = new ArrayList<>(Arrays.asList("partition-0","partition-1"));
+    SplitWorkUnitMessage splitMessage = new SplitWorkUnitMessage(workUnitId, laggingPartitions);
+
+    messageBuffer.publish(splitMessage);
+    fs.exists(path);
+    FileStatus[] fileStatus = fs.listStatus(path);
+    Assert.assertEquals(fileStatus.length, 1);
+    FSDataInputStream fis = fs.open(fileStatus[0].getPath());
+    DynamicWorkUnitMessage message = DynamicWorkUnitSerde.deserialize(ByteStreams.toByteArray(fis));
+
+    Assert.assertEquals(message.getWorkUnitId(), workUnitId);
+    Assert.assertEquals(((SplitWorkUnitMessage) message).getLaggingTopicPartitions().size(), 2);
+
+  }
+
+  @Test (dependsOnMethods={"testPublish"})
+  public void testSubscribe() throws IOException {
+    DynamicWorkUnitConsumer consumer = Mockito.mock(DynamicWorkUnitConsumer.class);
+
+    messageBuffer.subscribe(consumer);
+    try {
+      Thread.sleep(11000);
```

Comment:
Thread sleep in UT's should be avoided if possible or at least made really short to avoid having long CI. Can we shorten this value? 11 seconds seems a bit excessive. 

You can really cut down on this by having a short polling time

Reply:
Also, I think HDFS provides strong consistency (read-after-write) for the same client. Plus this is a local file system so risk is non existent w.r.t going to different name nodes.